src/Lean/Elab/Tactic/BuiltinTactic.lean

```diff
@@ -190,24 +190,26 @@ private def getOptRotation (stx : Syntax) : Nat :=
   let mvarIds ← getGoals
   let mut mvarIdsNew := #[]
   let mut abort := false
-  let mut mctxSaved ← getMCtx
   for mvarId in mvarIds do
     unless (← mvarId.isAssigned) do
       setGoals [mvarId]
+      let saved ← saveState
       abort ← Tactic.tryCatch
         (do
           evalTactic stx[1]
           pure abort)
         (fun ex => do
           if (← read).recover then
             logException ex
+            let msgLog ← Core.getMessageLog
+            saved.restore
+            Core.setMessageLog msgLog
```

Comment:
Should the trace / infotrees be restored too?

Reply:
Sorry, I meant "preserved from the failing tactic" rather than "restored from the saved state". Which one is the current behavior?

Reply:
So just to be super duper clear:

Preserved by default:
* traces
* `InfoTree`s

Not preserved by default:
* `Message`s

which is why this code manually un-restores the latter but not the former?

Reply:
Yes.

Traces are always preserved. Infotrees are by default preserved (there is a flag to restore them if wanted). MessageLogs are never preserved, hence the manual preservation.